Test 32-bit builds in CI and fix platform-specific code and tests

tautschnig opened the pull request:
Please review commit-by-commit.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Run 32-bit builds and tests as CI job

This is to detect hardwired 64-bit configuration or tests when those
should be platform-independent instead.

tautschnig committed:
Fix 32bit unit test build failure

GCC defines i386 as a macro (with value 1) on 32-bit x86.

TGWDB reviewed (commented):
Looks good, is this still in draft?

tautschnig commented:
> Looks good, is this still in draft?

Well, the tests are failing…

tautschnig committed:
Unit tests: do not hard-code 64-bit pointer width

We also support 32-bit architectures.

tautschnig committed:
Interpreter: Do not hard-code platform bit width

Not all systems have 64-bit addresses. Also fix an off-by-power-of-2
error in the comment.

tautschnig committed:
Make goto-inspect test patterns platform-agnostic

Not all platforms have 32-bit int, and the size of long int may be the
same as int.

tautschnig committed:
Test patterns: fix optional casts

Not all platforms require casts from int to long int as those may have
the same bit width.

tautschnig committed:
goto-harness regression tests: support non-64-bit platforms

The pointer width encoded in pre-built snapshots is 64 bits. Replace
that width at test runtime with the host system bit width.

tautschnig committed:
Arrays have a signed size

Fix tests and argc/argv constraints to account for the fact that our
arrays (at present) use signed sizes. We previously didn't always
consider the case that where sizeof(int) == sizeof(long), which
demonstrated overflows.

tautschnig committed:
Fix tests and contracts wrongly assuming sizeof(long) == sizeof(long long)

This isn't true for all platforms.

tautschnig committed:
Tests: use types that are smaller than long on all platforms

Some tests assumed that sizeof(int) < sizeof(long), which need not be
the case.

tautschnig committed:
Tests: do not hard-code 64 bits/8 bytes as pointer width

This wouldn't be true for all platforms.

tautschnig committed:
Tests: fix index typo

The comparison previously succeeded by chance on 64-bit platforms, but
didn't pass on ones that have a smaller pointer width.

tautschnig committed:
Properly type initial value of __CPROVER_max_malloc_size

Use a suffix to ensure the constant isn't treated as an int, but instead
matches the left-hand side's type.

tautschnig committed:
Goto analyzer: do not create untyped nil expression

We need to create a typed variant of top rather than relying on eval to
produce a typed expression. Tests involving pointer difference
spuriously passed for 64-bit platforms include a type cast in those
expressions. The abstract evaluation of said type cast resulted in typed
expressions despite the untyped nil input.

tautschnig committed:
Make cbmc-incr-smt2 test patterns platform-agnostic

Not all platforms have 32-bit int, and the size of long int may be the
same as int.

tautschnig committed:
cbmc-incr-smt2 tests: remove misguided assertion

There is no reason an integer couldn't have the same value as a pointer
(when converted to an integer type). This assertion might accidentally
hold on platforms where sizeof(int) != sizeof(int *) given our current
pointer model, but this is an implementation detail that we must not
rely on. The assertion certainly doesn't hold when sizeof(int) ==
sizeof(int *).

tautschnig committed:
Incremental SMT2: support cast from pointer to integer of different width

Do a zero extension, just like the propositional back-end does.

tautschnig committed:
Make memory-analyzer compatible with i386

We previously hard-coded x86_64 calling conventions for malloc.

tautschnig committed:
Contracts test: ensure return value access is in bounds

We ended up with an assumption that performed an out-of-bounds access as
there was no link between `size` and the return value in the contract.

tautschnig commented:
My apologies for producing such a large size of this bug-fixing PR: please do review commit-by-commit as each of them should be self-contained.

kroening commented:
This does remind me that we need to think about the type of the array size.

tautschnig commented:
> This does remind me that we need to think about the type of the array size.

Yes, we might be able to modify/partly revert the commit "[Arrays have a signed size](https://github.com/diffblue/cbmc/pull/7756/commits/767f44525c967f99f436d8069be28212e5c9ce42)" once/if we change that. (I tried to make sure that all changes related to array-size-signedness are contained in that one commit.)

TGWDB commented on src/memory-analyzer/gdb_api.cpp:
```diff
@@ -324,7 +333,7 @@ void gdb_apit::collect_malloc_calls()
   }
 
-  // now we can read the rax register to the the allocated memory address
+  // now we can read the rax register to the allocated memory address
```
Since we're here, if you are editing again before merge then maybe `rax` should be generalised since `i386` will only have `eax`.

tautschnig replied:
Done!

TGWDB commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -418,6 +418,52 @@ jobs:
       - name: Run tests
         run: cd build; ctest . -V -L CORE -j2
 
+  # This job takes approximately X to Y minutes
```
Reminder to update, something close to 60 by the looks of the last run.

tautschnig replied:
Done, thank you for your careful review!

martin-cs commented on regression/goto-analyzer/constant_propagation_01/test-vsd.desc:
```diff
@@ -3,7 +3,7 @@ main.c
-^Simplified:  assert: 1, assume: 0, goto: 1, assigns: 2, function calls: 0$
-^Unmodified:  assert: 0, assume: 0, goto: 0, assigns: 9, function calls: 2$
+^Simplified:  assert: 1, assume: 0, goto: 1, assigns: 1, function calls: 0$
+^Unmodified:  assert: 0, assume: 0, goto: 0, assigns: 10, function calls: 2$
```
I don't immediately see why this results in a simplification being skipped.  Do you have any intuition for what is going on here? Is this actually a VSD bug?

tautschnig replied:
The commit titled "Properly type initial value of __CPROVER_max_malloc_size" removes a type cast that the simplifier was able to take care of.

martin-cs commented on regression/goto-analyzer/constant_propagation_08/test-vsd.desc:
```diff
@@ -3,7 +3,7 @@ main.c
 ^SIGNAL=0$
-^Simplified:  assert: 1, assume: 0, goto: 1, assigns: 3, function calls: 0$
-^Unmodified:  assert: 0, assume: 0, goto: 2, assigns: 7, function calls: 2$
+^Simplified:  assert: 1, assume: 0, goto: 1, assigns: [12], function calls: 0$
```
Why the brackets?

martin-cs replied:
Ah, wait, is this platform dependent results?  If so this feels kinda bad.

tautschnig replied:
Fixed in #7912.

martin-cs commented on src/analyses/variable-sensitivity/abstract_pointer_object.cpp:
```diff
@@ -99,7 +99,7 @@ abstract_object_pointert abstract_pointer_objectt::eval_ptr_diff(
 {
   if(is_top() || operands[1]->is_top())
-    return environment.eval(nil_exprt(), ns);
+    return environment.abstract_object_factory(expr.type(), ns, true, false);
```
Good catch -- thank you.

martin-cs reviewed (approved):
Thoughts:

* Man I am glad I didn't have to do this!
* I guess this is what happens if things aren't in CI.
* I have concerns that the VSD regression test changes are actually uncovering some kind of bug.  Maybe we could have a look into this?

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -103,14 +103,14 @@ TEST_CASE("Value expr construction from smt.", "[core][smt2_incremental]")
     rowt{
       smt_bit_vector_constant_termt{0, 64},
-      from_integer(0, pointer_typet(void_type(), 64 /* bits */))},
+      from_integer(0, pointer_typet{empty_typet{}, 64 /* bits */})},
```
My apologies for nitpicking something which is already merged, but how does swapping from `void_type()` to `empty_typet{}` fix a 32-bit build problem? The `void_type()` function just returns an instance of `empty_typet{}`, and neither of these depend on a bit width, as far as I can see.

tautschnig replied:
Oh, yes, that indeed wasn't a complete fix: it was my attempt to get rid of `util/c_types.h`, for many of the types in there are by definition platform-dependent. But turns out that fully removing that include is impossible, because there also is use of `c_bool_typet` and `c_enum_typet`.

thomasspriggs replied:
Ah, so I could effectively think of this as a front-end de-coupling effort. There are many C front end specific types in the unit tests for the new SMT decision procedure because the decision procedure needs to implement support for them specifically. This is the case because these data structures find thier way all the way through cbmc to the decision procedure rather than being lowered to some kind of goto construct at the end of the processing of the front end. Another one for the list is `union_typet` which is defined in `util/c_types.h`, but which is sent to the decision procedure.

Whilst the decision procedure needs front-end specific functionality I would prefer to keep matching front-end specific types in the unit tests. That way we can remove the front end specifics from the decision procedure and the unit tests if and when we ever sufficiently decouple to stop them getting to the decision procedure.

tautschnig replied:
I entirely agree with your assessment. One day we should come clean on this and decide whether the types you mentioned are to be considered fundamental types, and therefore, should be moved out of `c_types.h`, or be lowered and not be ones to handle in back-ends.

tautschnig replied:
So, yes, my change in this file wasn't actually necessary, but probably doesn't need to be reverted either?

thomasspriggs replied:
I think it was somewhat easier to read as a void pointer, rather than an empty pointer. However I also don't think it is worth taking the time to revert. I questioned it as it didn't appear to match the commit message concerning 32 bit support, so I didn't initially understand why the change had been made. Thank you for taking the time to respond.
